### PR TITLE
add option to run MPIDriver in a single process

### PR DESCRIPTION
**Pull request by filipepgm**

### CODE REVIEW

**Comment by vloncar** [comment]
I would prefer if this functionality is implicit (i.e., when run with `mpirun -np 1` or even without MPI), rather than with a `--single` option. This way the `--single` option also need to be propagated to `mpiLAPI.py` if we want to use it in mpi_opt. The `MPISingleManager`s functionality looks like it could be gracefully rolled into existing `MPIManager` class. Did you investigate if the lines  [L75-L78](https://github.com/svalleco/mpi_learn/pull/22/commits/4e4929143c9edd246587e8bd03ca14bc3729c661#diff-0c0016dfb26f55df157a7eaf45c77075R75) of `MPISingleWorker` can be removed/simplified?

**Comment by vlimant** [comment]
this is very smart @filipepgm !
the purpose of this option is to be able to have the node=1 reference point for the performance scaling study for mpi-learn, in this respect it is not too much of an issue to have MPISingleManager. I however double @vloncar comment that we should try to roll into the existing MPIManager to limit the impact on mpi-opt

**Comment by filipepgm** [comment]
I have changed the code in order to keep the MPIManager as the interface, and removed the --single option.

I investigated L75-L78, but still have no good answers about it. I believe they could be removed, but doing so was breaking training. If the purpose is to keep the same functionality as with several processes, it may make sense to keep them.

File: mpi_learn/mpi/single_process.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+            epoch_metrics = epoch_metrics / float(i_batch+1)
+            l = self.model.get_logs( epoch_metrics )
+            self.update_history( l )
+
+            if self.stop_training:
+                break
+
+            self.validate()
+
+        print ("MPIWorker {0} signing off".format(self.ranks))
```
Comment (vloncar) [inline mpi_learn/mpi/single_process.py]: Forgot this one

File: mpi_learn/mpi/single_manager.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+from __future__ import division
+from .manager import MPIManager
+
+class MPISingleManager(MPIManager):
```
Comment (vloncar) [inline mpi_learn/mpi/single_manager.py]: Since you rolled this into `MPIManager`, this file can be removed.

**Comment by filipepgm** [comment]
Fixes applied

**Comment by vlimant** [comment]
is there a way to have validate defined only in one place, instead of having duplicated code to support ?

**Comment by filipepgm** [comment]
Changed to use the master's validate method.
Avoids code reuse, but may need restructuring if the validate method changes in any "complex" way

**Comment by vlimant** [comment]
looks good. @vloncar any final comments?

**Comment by vloncar** [comment]
Last commit introduces `check_parent` parameter to the `MPIWorker`'s constructor, which I don't think is really necessary (there is already plenty of parameters) and can be avoided by checking if the calling class is `MPISingleWorker` (like `if self.__class__.__name__ ...`). Also, since Filipe managed to trim `MPISingleWorker` to just 70 lines in a very elegant way, maybe the class itself can be moved into `process.py` so that we don't have a separate file.

**Pull request merged**